sklearn/tests/test_common.py

```diff
@@ -41,34 +43,57 @@ def test_all_estimator_no_base_class():
 
 
 def test_all_estimators():
-    # Test that estimators are default-constructible, cloneable
-    # and have working repr.
     estimators = all_estimators(include_meta_estimators=True)
 
     # Meta sanity-check to make sure that the estimator introspection runs
     # properly
     assert_greater(len(estimators), 0)
 
-    for name, Estimator in estimators:
-        # some can just not be sensibly default constructed
-        yield check_parameters_default_constructible, name, Estimator
 
+@pytest.mark.parametrize(
+        'name, Estimator',
+        all_estimators(include_meta_estimators=True)
+)
+def test_parameters_default_constructible(name, Estimator):
+    # Test that estimators are default-constructible
+    check_parameters_default_constructible(name, Estimator)
 
-def test_non_meta_estimators():
-    # input validation etc for non-meta estimators
-    estimators = all_estimators()
-    for name, Estimator in estimators:
+
+def _tested_non_meta_estimators():
+    for name, Estimator in all_estimators():
         if issubclass(Estimator, BiclusterMixin):
             continue
         if name.startswith("_"):
             continue
+        yield name, Estimator
+
+
+def _cartesian_product_checks(check_generator, estimators):
+    for name, Estimator in estimators:
         estimator = Estimator()
-        # check this on class
-        yield check_no_attributes_set_in_init, name, estimator
+        for check in check_generator(name, estimator):
+            yield name, Estimator, check
 
-        for check in _yield_all_checks(name, estimator):
-            set_checking_parameters(estimator)
-            yield check, name, estimator
+
+@pytest.mark.parametrize(
+        "name, Estimator, check",
+        _cartesian_product_checks(_yield_all_checks,
```

Comment:
Having multiple @parametrize will already produce the Cartesian product. How do we benefit from this explicit product generator?

Reply:
The name is a figure of speech; it's not rigorously a Cartesian product because for each estimator we get a list of checks that apply to it from `_yield_all_checks` not a predefined list of checks, so this cannot be writted as a two independent @parametrize decorators (or using `itertools.product`)..

Would the name `_combine_checks` or `_yield_checks_combination` be better ?

Reply:
Yes. Change the name. Thanks